File: ruby/lib/idkollen/transport.rb
# frozen_string_literal: true

require 'faraday'
require 'faraday/multipart'
require 'stringio'

module Idkollen
  USER_AGENT = "idkollen-client-ruby/#{Idkollen::VERSION}"

  class Transport
    def initialize(connection)
      @conn = connection
    end

    def post(path, body)
      handle(@conn.post(path) do |req|
        req.headers['Content-Type'] = 'application/json'
        req.body = body
      end)
    end

    def get(path)
      handle(@conn.get(path))
    end

    def get_raw(path)
      res = @conn.get(path)
      raise_for(res) unless success?(res)
      res.body.to_s
    end

    def delete(path)
      res = @conn.delete(path)
      raise_for(res) unless success?(res)
      nil
    end

    def post_multipart(path, data, filename, mime_type)
      io = StringIO.new(data)
      part = Faraday::Multipart::FilePart.new(io, mime_type, filename)
      handle(@conn.post(path, file: part))
    end

    private

    def handle(res)
      raise_for(res) unless success?(res)
      body = res.body
      return {} if body.nil? || body == ''
      body.is_a?(Hash) ? body : {}
    end

    def success?(res)
      res.status >= 200 && res.status < 300
    end

    def raise_for(res)
      message =
        case res.body
        when Hash then res.body['message'] || res.body.to_s
        when String then res.body
        else res.body.to_s
        end
      raise IdkollenError.new(res.status, message)
    rescue Faraday::Error => e
      raise IdkollenError.new(0, e.message)
    end
  end
end

File: swift/Sources/IdkollenClient/Transport.swift
import Foundation

internal final class Transport {
    private static let userAgent = "idkollen-client-swift/0.1.0"

    private let session: URLSession
    private let baseUrl: String
    private let authHeader: String
    let encoder: JSONEncoder
    let decoder: JSONDecoder

    init(session: URLSession, baseUrl: String, clientId: String, clientSecret: String) {
        self.session = session
        self.baseUrl = baseUrl.hasSuffix("/") ? String(baseUrl.dropLast()) : baseUrl
        let creds = Data("\(clientId):\(clientSecret)".utf8).base64EncodedString()
        self.authHeader = "Basic \(creds)"
        self.encoder = JSONEncoder()
        self.decoder = JSONDecoder()
    }

    func post<T: Decodable, B: Encodable>(_ path: String, body: B) async throws -> T {
        var req = buildRequest(method: "POST", path: path)
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = try encoder.encode(body)
        return try await send(req)
    }

    func get<T: Decodable>(_ path: String) async throws -> T {
        let req = buildRequest(method: "GET", path: path)
        return try await send(req)
    }

    func getRaw(_ path: String) async throws -> Data {
        let req = buildRequest(method: "GET", path: path)
        let (data, response) = try await runSession(req)
        try checkStatus(response: response, data: data)
        return data
    }

    func delete(_ path: String) async throws {
        let req = buildRequest(method: "DELETE", path: path)
        let (data, response) = try await runSession(req)
        try checkStatus(response: response, data: data)
    }

    func postMultipart<T: Decodable>(_ path: String, data: Data, filename: String, mimeType: String) async throws -> T {
        let boundary = UUID().uuidString
        var req = buildRequest(method: "POST", path: path)
        req.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")

        var body = Data()
        body.append("--\(boundary)\r\n".data(using: .utf8)!)
        body.append("Content-Disposition: form-data; name=\"file\"; filename=\"\(filename)\"\r\n".data(using: .utf8)!)
        body.append("Content-Type: \(mimeType)\r\n\r\n".data(using: .utf8)!)
        body.append(data)
        body.append("\r\n--\(boundary)--\r\n".data(using: .utf8)!)
        req.httpBody = body

        return try await send(req)
    }

    private func buildRequest(method: String, path: String) -> URLRequest {
        var req = URLRequest(url: URL(string: baseUrl + path)!)
        req.httpMethod = method
        req.setValue(authHeader, forHTTPHeaderField: "Authorization")
        req.setValue(Self.userAgent, forHTTPHeaderField: "User-Agent")
        return req
    }

    private func send<T: Decodable>(_ req: URLRequest) async throws -> T {
        let (data, response) = try await runSession(req)
        try checkStatus(response: response, data: data)
        if data.isEmpty {
            // Caller asked for T but body is empty. Decode an empty object if T allows it,
            // otherwise this throws — which is the right behaviour for non-optional T.
            return try decoder.decode(T.self, from: Data("{}".utf8))
        }
        return try decoder.decode(T.self, from: data)
    }

    private func runSession(_ req: URLRequest) async throws -> (Data, URLResponse) {
        do {
            return try await session.data(for: req)
        } catch {
            throw IdkollenError(statusCode: 0, message: error.localizedDescription)
        }
    }

    private func checkStatus(response: URLResponse, data: Data) throws {
        guard let http = response as? HTTPURLResponse else {
            throw IdkollenError(statusCode: 0, message: "Non-HTTP response")
        }
        guard !(200..<300).contains(http.statusCode) else { return }
        let message: String
        if let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
           let m = json["message"] as? String {
            message = m
        } else {
            message = String(data: data, encoding: .utf8) ?? ""
        }
        throw IdkollenError(statusCode: http.statusCode, message: message)
    }
}
